src/shogun/modelselection/NewGridSearch.cpp

```diff
@@ -0,0 +1,891 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#include <shogun/distance/Distance.h>
+#include <shogun/evaluation/CrossValidation.h>
+#include <shogun/kernel/Kernel.h>
+#include <shogun/modelselection/NewGridSearch.h>
+#include <shogun/util/factory.h>
+
+using namespace shogun;
+
+const std::string ParameterNode::delimiter = "::";
+
+ParameterNode::ParameterNode(CSGObject* model)
+{
+	SG_REF(model)
+	m_parent = model;
+	SG_SDEBUG("%s refcount: %d\n", model->get_name(), m_parent->ref_count())
+	for (auto const& param : m_parent->get_params())
+	{
+		// check the param is neither of type CLabels or CFeatures
+		if ((std::type_index(param.second->get_value().type_info()) !=
+		     std::type_index(typeid(CLabels*))) &&
+		    (std::type_index(param.second->get_value().type_info()) !=
+		     std::type_index(typeid(CFeatures*))))
+		{
+			if (auto* value = m_parent->get(param.first, std::nothrow))
+			{
+				auto node = new ParameterNode(value);
+				create_node(param.first, node);
+			}
+		}
+	}
+}
+
+void ParameterNode::create_node(const std::string& name, ParameterNode* obj)
+{
+	m_nodes[name].emplace_back(obj);
+}
+
+ParameterNode::ParameterNode(
+    CSGObject* model, ParameterProperties param_properties)
+{
+	SG_REF(model)
+	m_parent = model;
+
+	for (auto const& param : m_parent->get_params())
+	{
+		// check the param is neither of type CLabels or CFeatures
+		if ((std::type_index(param.second->get_value().type_info()) !=
+		     std::type_index(typeid(CLabels*))) &&
+		    (std::type_index(param.second->get_value().type_info()) !=
+		     std::type_index(typeid(CFeatures*))))
+		{
+			if (auto* value = m_parent->get(param.first, std::nothrow))
+			{
+				auto node = new ParameterNode(value);
+				create_node(param.first, node);
+			}
+		}
+	}
+
+	for (auto const& param : model->get_params(param_properties))
+	{
+		m_param_mapping.insert(
+		    std::make_pair(param.first, param.second->get_value()));
+	}
+}
+
+ParameterNode* ParameterNode::attach(
+    const std::string& param, const std::shared_ptr<ParameterNode>& node)
+{
+	return attach(param, node.get());
+}
+
+ParameterNode*
+ParameterNode::attach(const std::string& param, ParameterNode* node)
+{
+	auto* node_copy = new ParameterNode(*node);
+	if (!set_param_helper(param, node_copy))
+		SG_SERROR("Could not attach %s\n", param.c_str());
```

Comment:
would be good to see a reason if this is user-facing

Reply:
yea, I need to fix the error message a bit. It shouldn't reach this code in theory

Reply:
Then it could be a plain ASSERT in fact or? But that's minor